### PR TITLE
Add performance tester

### PR DESCRIPTION
**Pull request by zhkvia**
A tool for checking performance with tests description support.
See `clickhouse-performance-test --help` for additional information.

A test is written in XML and consists of:
- a name of a test
- how many times it should be processed
- tags/names for filtering
- preconditions for RAM size, cpu cache and data existing
- settings to apply to DBMS before running a test
- type of a query: `loop` or `once`
- stop criterions
- metrics for output
- substitutions for parametric queries

Main test (all metrics in JSON): https://gist.github.com/luc1ph3r/59aeeddb393e0a2c335f6477a34358dc

Test for mini-output (should be launched with `--lite` flag): https://gist.github.com/luc1ph3r/cae23cc9765f873c04706cb512bf0761

Test for average speed (default precision isn't big enough to stop. Kill it with SIGINT): https://gist.github.com/luc1ph3r/c019bda5e3dc6f472375cb0292128763

Test for preconditions (flushes CPU cache, checks for RAM size and data existence):
https://gist.github.com/luc1ph3r/619e953b9fca613a60711e0d5f525915

### CODE REVIEW

**Comment by robot-metrika-test** [comment]
Can one of the admins verify this patch?

File: dbms/src/Client/PerformanceTest.cpp [hunk 1]
```diff
@@ -0,0 +1,1556 @@
...
+
+#include "InterruptListener.h"
+
+/** Tests launcher for ClickHouse.
+  * The tool walks through given or default folder in order to find files with
+  * tests' descriptions and launches it.
+  */
+namespace FS = boost::filesystem;
+using String = std::string;
+const String FOUR_SPACES = "    ";
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: It is better to make it `static`

File: dbms/src/Client/PerformanceTest.cpp [hunk 2]
```diff
@@ -0,0 +1,1556 @@
...
+    extern const int POCO_EXCEPTION;
+    extern const int STD_EXCEPTION;
+    extern const int UNKNOWN_EXCEPTION;
+}
+
+static String pad(size_t padding) {
+    return String(padding * 4, ' ');
+}
+
+class JSONString
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: A comment is required

File: dbms/src/Client/PerformanceTest.cpp [hunk 3]
```diff
@@ -0,0 +1,1556 @@
...
+    std::vector<Config> tests_configurations;
+
+    Strings tests_tags;
+    Strings skip_tags;
+    Strings tests_names;
+    Strings skip_names;
+    Strings tests_names_regexp;
+    Strings skip_names_regexp;
+
+    #define incFulfilledCriterions(index, CRITERION)                                                                        \
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: Macros should not looks like a regular function.
It should has CAPS name, be declared directly before it usage and be `#undef`ined after last usage.

File: dbms/src/Client/PerformanceTest.cpp [hunk 4]
```diff
@@ -0,0 +1,1556 @@
...
+                    return false;
+                }
+
+            if (precondition == "ram_size")
+            {
+                struct sysinfo *system_information = new struct sysinfo();
+                if (sysinfo(system_information))
+                {
+                    std::cerr << "Failed to check system RAM size" << std::endl;
+                    delete system_information;
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: `return false` ?

File: dbms/src/Client/PerformanceTest.cpp [hunk 5]
```diff
@@ -0,0 +1,1556 @@
...
+
+        if (tests_configurations.size())
+        {
+            Strings outputs;
+
+            for (auto & test_config : tests_configurations)
+            {
+                if (!checkPreconditions(test_config))
+                {
+                    std::cerr << "Preconditions are not fulfilled for test \"" + test_config->getString("name", "") + "\"";
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: What we will see in output in this case? Will it contain a note signalling about failed prerequisites?

File: dbms/src/Client/PerformanceTest.cpp [hunk 6]
```diff
@@ -0,0 +1,1556 @@
...
+
+            String value = config->getString(key + "." + setup);
+            if (value.empty())
+                value = "true";
+
+            settings_to_apply[setup] = value;
+        }
+    }
+
+    String runTest(Config & test_config)
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: Comment is required (what does it return)?

File: dbms/src/Client/PerformanceTest.cpp [hunk 7]
```diff
@@ -0,0 +1,1556 @@
...
+
+        if (!main_metric.empty())
+        {
+            if (std::find(metrics.begin(), metrics.end(), main_metric) == metrics.end())
+                metrics.push_back(main_metric);
+        }
+        else
+        {
+            if (lite_output)
+                throw DB::Exception("Specify main_metric for lite output", 1);
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: Use meaningful error code (second argument) from `ErrorCodes.cpp`

File: dbms/src/Client/PerformanceTest.cpp [hunk 8]
```diff
@@ -0,0 +1,1556 @@
...
+        }
+
+        if (interrupt_listener.check())
+        {
+            gotSIGINT = true;
+            stream.cancel();
+        }
+    }
+
+    void constructSubstitutions(AbstractConfig & substitutions_view, StringToVector & substitutions)
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: Maybe `const AbstractConfig & substitutions_view`?
And a comment would be helpful.

File: dbms/src/Client/PerformanceTest.cpp [hunk 9]
```diff
@@ -0,0 +1,1556 @@
...
+                StringToVector::iterator next_it = substitutions_left;
+                ++next_it;
+
+                runThroughAllOptionsAndPush(next_it, substitutions_right, query, queries, substitutions_map);
+            }
+        }
+    }
+
+public:
+    String constructTotalInfo(Strings metrics)
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: A comment is required

File: dbms/src/Client/PerformanceTest.cpp [hunk 10]
```diff
@@ -0,0 +1,1556 @@
...
+        }
+
+        return output;
+    }
+};
+}
+
+static void getFilesFromDir(const FS::path & dir, std::vector<String> & input_files)
+{
+    if (dir.extension().string() == ".xml")
```
Comment (ludv1x) [inline dbms/src/Client/PerformanceTest.cpp]: Looks like excessive check

**Comment by orivej** [comment]
After this change ClickHouse programs fail to build when configured with `-DUNBUNDLED=ON -DUSE_STATIC_LIBRARIES=OFF` because `libclickhouse.so` is not linked with `boost_filesystem`.

```
g++   -D_GLIBCXX_USE_CXX11_ABI=1 -pipe -msse4.1 -msse4.2 -mpopcnt -std=gnu++1z  -fno-omit-frame-pointer -Wall -Wnon-virtual-dtor  -Werror -Os -DNDEBUG  -static-libgcc -static-libstdc++  -D_GLIBCXX_USE_CXX11_ABI=1 @CMakeFiles/iotest.rsp  -o utils/iotest/iotest
dbms/libclickhouse.so.1.1.54233: undefined reference to `boost::program_options::abstract_variables_map::operator[](std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
dbms/libclickhouse.so.1.1.54233: undefined reference to `vtable for boost::program_options::error_with_option_name'
dbms/libclickhouse.so.1.1.54233: undefined reference to `boost::program_options::variables_map::variables_map()'
dbms/libclickhouse.so.1.1.54233: undefined reference to `boost::program_options::error_with_option_name::what() const'
dbms/libclickhouse.so.1.1.54233: undefined reference to `typeinfo for boost::program_options::error_with_option_name'
…
```

https://github.com/orivej/ClickHouse/commit/19200a3e25198e4965261de4c49174d4433e8804 fixes this, but I do not know if it is the right fix.

**Comment by orivej** [comment]
This happens because `add_headers_and_sources(dbms src/Client)` in `dbms/CMakeLists.txt` pulls `dbms/src/Client/PerformanceTest.cpp` into `libclickhouse.so`, but does not add link dependency on `boost_filesystem`. Configuring with `-DUSE_STATIC_LIBRARIES=OFF` is sufficient to reproduce the issue (i.e. `-DUNBUNDLED=ON` is not necessary).

**Comment by alexey-milovidov** [comment]
Thanks for investigation! Is this Ok? https://github.com/yandex/ClickHouse/commit/0fcb6300a2601edbf98a2b32d170dc48d032472a

**Comment by orivej** [comment]
Yes, now it's fine.

**Pull request merged**